### PR TITLE
make: enable selection of periph drivers via USEMODULE

### PR DESCRIPTION
**Pull request by haukepetersen**
So far, we are compiling all available peripheral drivers for a CPU, independent if they are really used or not. This PR enables us to treat peripheral drivers just as ordinary modules (e.g. select them via `USEMODULE += periph_gpio`, etc.).

I re-ordered some includes in the main `Makefile.include`, would be nice if somebody could re-check this change! @gebart, @Kijewski maybe?

This PR brings some advantages:
- we save compile time
- we can check automatically during compile time, if a board/cpu has all the peripheral drivers needed -> no need to explicitly state the required features in an application anymore...
- device drivers now state explicitly, what kind of peripherals they need. This could be used for automatic extraction of a board-to-driver compatibility matrix or the like

Outlook: Next step is to look at the test application and scripts and remove the explicit `FEATURES_REQUIRED` definitions from their Makefiles...

### CODE REVIEW

File: drivers/Makefile.dep [hunk 1]
```diff
@@ -0,0 +1,163 @@
...
+
+ifneq (,$(filter at86rf231,$(USEMODULE)))
+  USEMODULE += periph_gpio
+  USEMODULE += periph_spi
+  USEMODULE += netdev_802154
+  USEMODULE += ieee802154
+endif
+
+ifneq (,$(filter cc110x%,$(USEMODULE)))
+  USEMODULE += protocol_multiplex
```
Comment (gebart) [inline drivers/Makefile.dep]: don't the `cc***` use any periph stuff?

**Comment by gebart** [comment]
I think Mulle would need `USEMODULE += periph_gpio` because it controls on-board LEDs and power switches using the periph_gpio driver.

File: drivers/Makefile.dep (hunk 1 above)
Comment (haukepetersen) [inline drivers/Makefile.dep]: nope, these are old (deprecated) drivers. The SPI/GPIO stuff of this drivers is implemented in a board specific part of the driver for some of the boards... These drivers are anyway subject to cleanup soon and will be removed.

**Comment by haukepetersen** [comment]
The thing is, that all Freescale CPUs do still build all the peripherals anyway, as the structure for these boards differs from the structure of the other CPUs. I would actually touch them in a follow-up PR...

**Comment by haukepetersen** [comment]
rebased.

**Comment by PeterKietzmann** [comment]
@kaspar030 did you already start your review? For #3538 it would be cool to have this one merged soon. At the first glance the PR looks ok to me, but I didn't read the changes that detailed.

**Comment by haukepetersen** [comment]
@kaspar030 @PeterKietzmann could you review this? Would be nice to have it off the table.

Also rebased.

File: drivers/Makefile.dep (hunk 1 above)
Comment (kaspar030) [inline drivers/Makefile.dep]: kick the line out anyways, there's no cc110x in master anymore. :)

**Comment by kaspar030** [comment]
Looking at it now, is there a need for all "periph" *_MODULEs to also be _FEATURES?

**Comment by haukepetersen** [comment]
Don't know if I understand the questions correctly: do you mean if we still need the features_provided declarations? Then the answer is yes.

File: boards/saml21-xpro/Makefile.features [hunk 2]
```diff
@@ -1,6 +1,7 @@
+FEATURES_PROVIDED += cpp
+FEATURES_PROVIDED += periph_timer
 FEATURES_PROVIDED += periph_gpio
 FEATURES_PROVIDED += periph_spi
-FEATURES_PROVIDED += cpp
 FEATURES_PROVIDED += periph_rtt
 FEATURES_PROVIDED += periph_timer
```
Comment (kaspar030) [inline boards/saml21-xpro/Makefile.features]: timer in there two times

File: drivers/Makefile.dep [hunk 3]
```diff
@@ -0,0 +1,163 @@
...
+  USEMODULE += periph_gpio
+  USEMODULE += periph_spi
+endif
+
+ifneq (,$(filter at86rf231,$(USEMODULE)))
+  USEMODULE += periph_gpio
+  USEMODULE += periph_spi
+endif
+
+ifneq (,$(filter cc110x,$(USEMODULE)))
```
Comment (kaspar030) [inline drivers/Makefile.dep]: this can go

File: drivers/Makefile.dep [hunk 4]
```diff
@@ -0,0 +1,163 @@
...
+endif
+
+ifneq (,$(filter xbee,$(USEMODULE)))
+  USEMODULE += periph_gpio
+  USEMODULE += periph_uart
+  USEMODULE += ng_ieee802154
+endif
+
+
+# Deprecated drivers
```
Comment (kaspar030) [inline drivers/Makefile.dep]: they can also go

**Comment by haukepetersen** [comment]
hm, the new `periph_common` clashes with this PR, as this PR does automatic parsing of everything that starts with `periph_`...

**Comment by kaspar030** [comment]
Do you want to rename that to "common_periph"?

**Comment by haukepetersen** [comment]
just did.

**Comment by haukepetersen** [comment]
Can you look over the changes once more? I will squash once you give me your ok.

File: drivers/Makefile.dep [hunk 5]
```diff
@@ -0,0 +1,131 @@
+# driver dependencies (in alphabetical order)
```
Comment (kaspar030) [inline drivers/Makefile.dep]: I like!

File: drivers/at86rf2xx/at86rf2xx.c [hunk 6]
```diff
@@ -88,7 +88,7 @@ int at86rf2xx_init(at86rf2xx_t *dev, spi_t spi, spi_speed_t spi_speed,
 
 void at86rf2xx_reset(at86rf2xx_t *dev)
 {
-#if CPUID_ID_LEN
+#if MODULE_PERIPH_CPUID
```
Comment (kaspar030) [inline drivers/at86rf2xx/at86rf2xx.c]: Usually we "#ifdef MODULE_*". This probably doesn't work as expected.

File: drivers/at86rf2xx/at86rf2xx.c [hunk 7]
```diff
@@ -107,7 +107,7 @@ void at86rf2xx_reset(at86rf2xx_t *dev)
     dev->seq_nr = 0;
     dev->options = 0;
     /* set short and long address */
-#if CPUID_ID_LEN
+#if MODULE_PERIPH_CPUID
```
Comment (kaspar030) [inline drivers/at86rf2xx/at86rf2xx.c]: #ifdef

File: drivers/kw2xrf/kw2xrf.c [hunk 8]
```diff
@@ -384,7 +384,7 @@ int kw2xrf_init(kw2xrf_t *dev, spi_t spi, spi_speed_t spi_speed,
     uint8_t reg = 0;
     uint8_t tmp[2];
     kw2xrf_gpio_int = int_pin;
-#if CPUID_ID_LEN
+#if MODULE_PERIPH_CPUID
```
Comment (kaspar030) [inline drivers/kw2xrf/kw2xrf.c]: #ifdef

File: drivers/kw2xrf/kw2xrf.c [hunk 9]
```diff
@@ -411,7 +411,7 @@ int kw2xrf_init(kw2xrf_t *dev, spi_t spi, spi_speed_t spi_speed,
     dev->proto = KW2XRF_DEFAULT_PROTOCOL;
     dev->option = 0;
 
-#if CPUID_ID_LEN
+#if MODULE_PERIPH_CPUID
```
Comment (kaspar030) [inline drivers/kw2xrf/kw2xrf.c]: ...

File: sys/net/application_layer/ng_zep/ng_zep.c [hunk 10]
```diff
@@ -92,7 +92,7 @@ static uint16_t _calc_fcs(uint16_t fcs, const uint8_t *frame, uint8_t frame_len)
 kernel_pid_t ng_zep_init(ng_zep_t *dev, uint16_t src_port, ipv6_addr_t *dst,
                          uint16_t dst_port)
 {
-#if CPUID_ID_LEN
+#if MODULE_PERIPH_CPUID
```
Comment (kaspar030) [inline sys/net/application_layer/ng_zep/ng_zep.c]: ...

File: sys/net/application_layer/ng_zep/ng_zep.c [hunk 11]
```diff
@@ -121,7 +121,7 @@ kernel_pid_t ng_zep_init(ng_zep_t *dev, uint16_t src_port, ipv6_addr_t *dst,
     dev->chan = NG_ZEP_DEFAULT_CHANNEL;
     dev->pan = byteorder_btols(byteorder_htons(NG_ZEP_DEFAULT_PANID));
     dev->flags = NG_ZEP_FLAGS_USE_SRC_PAN;
-#if CPUID_ID_LEN
+#if MODULE_PERIPH_CPUID
```
Comment (kaspar030) [inline sys/net/application_layer/ng_zep/ng_zep.c]: ...

**Comment by kaspar030** [comment]
Looking good apart the "#if MODULE*".

File: drivers/at86rf2xx/at86rf2xx.c (hunk 6 above)
Comment (haukepetersen) [inline drivers/at86rf2xx/at86rf2xx.c]: true, that one sipped through...

**Comment by haukepetersen** [comment]
fixed

**Comment by haukepetersen** [comment]
rebased. @kaspar030: ping?

File: drivers/Makefile.dep [hunk 12]
```diff
@@ -0,0 +1,131 @@
...
+endif
+
+ifneq (,$(filter tmp006,$(USEMODULE)))
+  USEMODULE += periph_i2c
+endif
+
+ifneq (,$(filter xbee,$(USEMODULE)))
+  USEMODULE += periph_gpio
+  USEMODULE += periph_uart
+  USEMODULE += ng_ieee802154
```
Comment (PeterKietzmann) [inline drivers/Makefile.dep]: I think it's not "ng_" any more. Right?

File: sys/net/gnrc/application_layer/zep/gnrc_zep.c [hunk 13]
```diff
@@ -120,8 +120,9 @@ kernel_pid_t gnrc_zep_init(gnrc_zep_t *dev, uint16_t src_port, ipv6_addr_t *dst,
     dev->driver = (gnrc_netdev_driver_t *)&_zep_driver;
     dev->chan = GNRC_ZEP_DEFAULT_CHANNEL;
     dev->pan = byteorder_btols(byteorder_htons(GNRC_ZEP_DEFAULT_PANID));
-    dev->flags = GNRC_ZEP_FLAGS_USE_SRC_PAN;
-#if CPUID_ID_LEN
+    dev->flags = GNRC_ZEP_FLAGS_USE_SRC_PAN
```
Comment (PeterKietzmann) [inline sys/net/gnrc/application_layer/zep/gnrc_zep.c]: No semicolon?

**Comment by PeterKietzmann** [comment]
There are several build issues according to Travis

**Comment by haukepetersen** [comment]
oh, will look into it now

**Comment by haukepetersen** [comment]
fixed obvious errors and look into travis bugs now.

**Comment by haukepetersen** [comment]
Travis things seemed unrelated to me, let's see what the current run tells us.

File: drivers/Makefile.dep [hunk 14]
```diff
@@ -0,0 +1,131 @@
...
+ifneq (,$(filter ina220,$(USEMODULE)))
+  USEMODULE += periph_i2c
+endif
+
+ifneq (,$(filter isl29020,$(USEMODULE)))
+  USEMODULE += periph_i2c
+endif
+
+ifneq (,$(filter isl29125,$(USEMODULE)))
+  USEMODULE += periph_gpio
```
Comment (PeterKietzmann) [inline drivers/Makefile.dep]: This driver is based on i2c. Don't you need to add the `USEMODULE += periph_i2c` line here?

File: drivers/Makefile.dep [hunk 15]
```diff
@@ -0,0 +1,131 @@
...
+
+ifneq (,$(filter tmp006,$(USEMODULE)))
+  USEMODULE += periph_i2c
+endif
+
+ifneq (,$(filter xbee,$(USEMODULE)))
+  USEMODULE += periph_gpio
+  USEMODULE += periph_uart
+  USEMODULE += ieee802154
+endif
```
Comment (PeterKietzmann) [inline drivers/Makefile.dep]: Are you sure you listed all drivers here? I didn't check, but what happened to `hih6130` for example?

File: boards/yunjia-nrf51822/Makefile.features [hunk 16]
```diff
@@ -1,5 +1,6 @@
 FEATURES_PROVIDED += cpp
 FEATURES_PROVIDED += radio_nrfmin
+FEATURES_PROVIDED += periph_timer
```
Comment (PeterKietzmann) [inline boards/yunjia-nrf51822/Makefile.features]: I think you should also add this line to the other nrf51 based boards, right?

File: drivers/Makefile.dep [hunk 17]
```diff
@@ -0,0 +1,131 @@
...
+ifneq (,$(filter at86rf2%,$(USEMODULE)))
+  USEMODULE += at86rf2xx
+  USEMODULE += ieee802154
+  USEMODULE += periph_gpio
+  USEMODULE += periph_spi
+  ifneq (,$(filter periph_cpuid,$(FEATURES_PROVIDED)))
+    USEMODULE += periph_cpuid
+  endif
+endif
+
```
Comment (PeterKietzmann) [inline drivers/Makefile.dep]: ... `undefined reference to `cpuid_get'`. Don't you need to add an entry for the nrfmin driver here or somewhere?

**Comment by PeterKietzmann** [comment]
I don't know where this comes from, but some applications don't compile because it seems that a USEMODULE is missing. E.g. now you can't compile `periph_adc` for stm boards. Adding `USEMODULE += periph_adc` to the tests Makefile, it works

**Comment by PeterKietzmann** [comment]
>Travis things seemed unrelated to me, let's see what the current run tells us

@haukepetersen am I mistaken somewhere? For me, the build fails really don't seem unrelated to the PR. Please adapt some of my proposed changes. I hope the number of fails will decrease and then I'll have a look at your PR again. Ok?

**Comment by haukepetersen** [comment]
I think you are right, I mussed have messed up something in the build system... I look into it as soon as I find the time...

**Comment by haukepetersen** [comment]
remove the hack'n'ack label, as this PR still needs work

**Comment by haukepetersen** [comment]
ok won't get this working anytime soon (or ever?), so I will close this PR for now and open a new one whenever (if ever) I make progress...

**Pull request closed without merging**